### PR TITLE
Added the hammerhead.EVENTS.unhandledRejection event (close #1247)

### LINKED ISSUE #1247: Add the hammerhead.EVENTS.unhandledRejection event
see https://developer.mozilla.org/ru/docs/Web/API/WindowEventHandlers/onunhandledrejection

It's necessary for https://github.com/DevExpress/testcafe/issues/1473

### PR DESCRIPTION
**Pull request by LavrovArtem**
/cc @miherlosev @AlexanderMoskovkin

### CODE REVIEW

File: src/client/sandbox/event/message.js [hunk 1]
```diff
@@ -207,8 +207,7 @@ export default class MessageSandbox extends SandboxBase {
                     // trying to call the 'isCrossDomainWindows' function, so we have to wrap it in 'try/catch'.
                     try {
                         targetWindow[this.RECEIVE_MSG_FN]({
-                            // NOTE: Cloning a message to prevent this modification.
```
Comment (LavrovArtem) [inline src/client/sandbox/event/message.js]: Controversial place. Do we need this cloning?
/cc @miherlosev @AlexanderMoskovkin

Comment (AlexanderMoskovkin) [inline src/client/sandbox/event/message.js]: Are all tests are ok with this? I suppose it should be something sensitive since it was especially commented

Comment (LavrovArtem) [inline src/client/sandbox/event/message.js]: I removed the test which checks that message was cloned.
https://github.com/DevExpress/testcafe-hammerhead/pull/1249/files#diff-366a593bdf9d1d4547600cef78eedf11L175

File: test/client/fixtures/sandbox/node/window-test.js [hunk 2]
```diff
@@ -16,6 +17,55 @@ test('window.onerror setter/getter', function () {
     strictEqual(getProperty(window, 'onerror'), handler);
 });
 
+if (featureDetection.hasUnhandledRejectionEvent) {
+    module('window.onunhandledrejection', function () {
+        asyncTest('must be overriden (GH-1247)', function () {
```
Comment (miherlosev) [inline test/client/fixtures/sandbox/node/window-test.js]: Add test for case then page has handler for `window.onunhandledrejection`. In this case we should not raise error.

File: src/client/sandbox/event/message.js (hunk 1 above)
Comment (miherlosev) [inline src/client/sandbox/event/message.js]: Revert back message cloning and add more detailed comment

**Comment by LavrovArtem** [comment]
@testcafe-build-bot \retest

File: src/client/sandbox/node/window.js [hunk 3]
```diff
@@ -16,44 +16,57 @@ import constructorIsCalledWithoutNewKeyword from '../../utils/constructor-is-cal
...
+        nativeMethods.windowRemoveEventListener.call(window, 'unhandledrejection', this);
+        nativeMethods.windowAddEventListener.call(window, 'unhandledrejection', this);
+    }
 
-                this.messageSandbox.sendServiceMsg({
-                    msg,
-                    pageUrl,
+    handleEvent (event) {
+        if (event.type === 'unhandledrejection' && !event.defaultPrevented) {
+            const reason = event.reason;
```
Comment (miherlosev) [inline src/client/sandbox/node/window.js]: Move to the separate function `WindowSandbox.formatUnhandledRejectionReason (eventReason)` and covers it by client tests

File: src/client/sandbox/node/window.js [hunk 4]
```diff
@@ -16,44 +16,57 @@ import constructorIsCalledWithoutNewKeyword from '../../utils/constructor-is-cal
...
+            }
+            else
+                this.emit(type, { msg, pageUrl });
+        }
+    }
 
-                    inIframe: true
-                });
+    _reattachUnhandledRejectionHandler (window) {
+        nativeMethods.windowRemoveEventListener.call(window, 'unhandledrejection', this);
```
Comment (miherlosev) [inline src/client/sandbox/node/window.js]: It is difficult to understand. 
Change to `nativeMethods.windowRemoveEventListener.call(window, 'unhandledrejection', this.handleEvent)`

Comment (LavrovArtem) [inline src/client/sandbox/node/window.js]: It calls this.handleEvent without WindowSandbox context

**Comment by LavrovArtem** [comment]
@testcafe-build-bot \retest

**Comment by LavrovArtem** [comment]
@testcafe-build-bot \retest

**Comment by LavrovArtem** [comment]
@testcafe-build-bot \retest

**Comment by LavrovArtem** [comment]
@testcafe-build-bot \retest

**Comment by LavrovArtem** [comment]
@testcafe-build-bot \retest

File: src/client/sandbox/node/window.js [hunk 5]
```diff
@@ -16,44 +16,62 @@ import constructorIsCalledWithoutNewKeyword from '../../utils/constructor-is-cal
...
-                    cmd: this.UNCAUGHT_JS_ERROR_EVENT
-                }, window.top);
-            }
-            else {
-                this.emit(this.UNCAUGHT_JS_ERROR_EVENT, {
-                    msg,
-                    pageUrl
-                });
-            }
+        if (reason === null || reason === void 0 || reasonType === 'number' || reasonType === 'boolean')
```
Comment (miherlosev) [inline src/client/sandbox/node/window.js]: Let's sturcture as
```js
if (utils.isPrimitiveType (reason))
    reasonStr = String(reason)
else if (reason instanceof Error)
   reasonStr = reason.message;
else
  reasonStr = nativeMethods.objectToString.call(reason);
```

where 
```js
isPrimitiveType (obj) {
  var objType = typeof obj;

 return objType !== 'object' && objType !== 'function';
}
```

File: test/client/fixtures/sandbox/node/window-test.js [hunk 6]
```diff
@@ -16,6 +17,144 @@ test('window.onerror setter/getter', function () {
...
+                    return testMsg(void 0);
+                })
+                .then(function (msg) {
+                    strictEqual(msg, 'undefined');
+                    return testMsg(function () {
+                    });
+                })
+                .then(function (msg) {
+                    strictEqual(msg, '[object Function]');
+                });
```
Comment (miherlosev) [inline test/client/fixtures/sandbox/node/window-test.js]: Add test for `Symbol` type.
Reorder check as : null, undefined, string, number, boolean, symbol, object, function

**Pull request merged**